---
title: "Group Project"
authors: "Gabrielle McCabe, Ethan Cohen, Jenna Kim, Kevin Saroya, Aimee Watts"
date: "10/04/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Biological Questions
1. Do the belly button microbiome differ between the two populations from which samples were taken from?
2. What is the evolutionary relationship between the different species found in one sample (one belly button)?

# Libraries
```{r}
library(ggplot2)
library(ape)
library(ggtree)
library(vegan)
library(seqinr)
library(reshape2)
```
# Question 1

## Setup Sample Information

Import the sample information.
```{r}
Samples <- read.csv("data/sample_info.csv")
```

## Setup OTU table

Import OTU table.
```{r}
OTU_table <- read.delim("data/OTU_file.txt", header = T, row.names = "X.OTU.ID")
```

Remove taxonomy column from OTU table.
```{r}
OTU_data <- OTU_table[, -c(1, ncol(OTU_table))]
```

Calculate the total sequences in the dataset.
```{r}
x <- rowSums(OTU_data)
sum(x)
```

With 24,000 reads there are most likely going to be some reads that are only contamination. Therefore, to remove these and to reduce the dataset slightly we will remove any OTUs that do not have more than one sequence in more than one sample.
```{r}
drop <- rowSums(OTU_data) < 2
sum(drop)  # The number of sequences being removed.
OTU_red <- OTU_data[!drop, ]
```

Transpose the table so that the species are across the top and the samples are along the side.
```{r}
OTU_red[1:3, 1:3]  # Preview the layout of the original data.frame. 
OTUs <- as.data.frame(t(OTU_red))
OTUs[1:3, 1:3]  # Ensure transpose worked.
```

## Analysis
### Binary Method

Change OTUs data.frame into binary data by changing the read counts in each sequence to 1 or 0.
```{r}
OTU_bin <- OTUs
OTU_bin[OTU_bin > 0] <- 1
```

Calculate the pairwise distance of the binary matrix.
```{r}
OTU_bin_dist <- dist(OTU_bin, method = "binary")
```

Build the binary neighbour-joining tree, annotate it using the sample information, and output it as a pdf.
```{r}
OTU_bin_tree <- nj(OTU_bin_dist)
ggtree(OTU_bin_tree, layout = "rectangular") %<+% Samples +
  geom_tiplab(aes(colour = collection)) +
  theme(legend.position = "right")

pdf(width = 16, height = 20, "Binary_Tree.pdf")
ggtree(OTU_bin_tree, layout = "rectangular") %<+% Samples +
  geom_tiplab(aes(colour = collection)) +
  theme(legend.position = "right")
dev.off()
```

### Euclidean Method

Calculate the euclidean distance.
```{r}
OTU_euc_dist <- dist(OTUs, method = "euclidean") 
```

Build the euclidean neighbour-joining tree, annotate it using the sample information, and output it as a pdf.
```{r}
OTU_euc_tree <- nj(OTU_euc_dist)               
ggtree(OTU_euc_tree, layout = "rectangular") %<+% Samples +
  geom_tiplab(aes(colour = collection)) +
  theme(legend.position = "right")

pdf(width = 16, height = 20, "Euclidean_Tree.pdf")
ggtree(OTU_euc_tree, layout = "rectangular") %<+% Samples +
  geom_tiplab(aes(colour = collection)) +
  theme(legend.position = "right")
dev.off()
```

### Bray-Curtis Dissimilarity Method

Calculate the Bray-Curtis dissimilarity.
```{r}
OTU_bc_dist <- vegdist(OTUs, method = "bray", binary = T)
```

Build the Bray-Curtis dissimilarity neighbour-joining tree, annotate it using the sample information, and output it as a pdf.
```{r}
OTU_bc_tree <- nj(OTU_bc_dist)
ggtree(OTU_bc_tree, layout = "rectangular") %<+% Samples +
  geom_tiplab(aes(colour = collection)) +
  theme(legend.position = "right")

pdf(width = 16, height = 20, "Bray_Curtis_Tree.pdf")
ggtree(OTU_bc_tree, layout = "rectangular") %<+% Samples +
  geom_tiplab(aes(colour = collection)) +
  theme(legend.position = "right")
dev.off()
```

### Non-Metric Multidimensional Scaling (NMDS)

Perform the model.
```{r}
set.seed(13)
NMDSdat <- metaMDS(OTU_bc_dist, k = 2, trymax = 100)
```

Create data for plotting.
```{r}
PDat <- data.frame(NMDS1 = NMDSdat$points[ , 1],
                   NMDS2 = NMDSdat$points[ , 2],
                   sample = row.names(OTUs))
```

Add species labels.
```{r}
PDat <- merge(PDat, Samples, by = "sample", all.x = T, all.y = F)
```

Plot the NMDS.
```{r}
qplot(x = NMDS1, NMDS2, colour = collection, alpha = I(0.6), data = PDat) +
  theme_bw()
```

# Question 2

## Data setup

Importing the Fasta file and selecting a single sample, B1285, in order to determine the evolutionary diversity among the bacteria and archaea within that single sample.
A for loop was utilized to gather all the sequence reads which pertained to the individual sample collected (B1285)
```{r}
myFasta <- read.fasta(file = "data/raw_seqs_BB.fna", seqtype = "AA", as.string = TRUE, set.attributes = FALSE)
nam <- names(myFasta)

indexes <- grep("B1285", nam)

sub <- rep(NA, 1170)

for (i in 1:length(indexes)){
  sub[i] <- myFasta[indexes[i]]
}
```

Creating a dataframe using index numbers as IDs and the sequence data from the myFasta file pasted in the seq column. A new object 'dna' was created
using the sapply function to seperate each base pair into seperate columns. Following, the names function was utilized to re-name all row names to their corresponding indices from the orignial file read into the myFasta file.
```{r}
df <- data.frame(ID = as.factor(indexes), Seq = paste(sub), stringsAsFactors = FALSE)
dna <- sapply(df$Seq, strsplit, split = "")
names(dna) <- paste(1:nrow(df),df$ID,sep="_")
dna_bin<-as.DNAbin(dna)
```

## Alignment

The 'dna' file was converted to a DNAbin object which can be acted upon the muscle to align the sequence data. 
```{r}
dna_align <- muscle(dna_bin,quiet=F)
```

## Inspect the alignment

Visual of the alignment of all sequences at all sites. There is too much going on to determine anything so we will zoom in on the beginning, middle, and end of the alignment for the first half of the sequences and then the second half.
```{r}
checkAlignment(dna_align, what = 1)
```

The beginning of the alignment for the first half of the sequences. Note the large gaps right before and after site 80.
```{r}
checkAlignment(dna_align[1:585, 1:150], what = 1)
```

The beginning of the alignment for the second half of the sequences. Note the large gaps right before and after site 80.
```{r}
checkAlignment(dna_align[586:1170, 1:150], what = 1)
```

The middle of the alignment for the first half of the sequences. Note the large gap before site 10.
```{r}
checkAlignment(dna_align[1:585, 151:300], what = 1)
```

The middle of the alignment for the second half of the sequences. Note the large gap before site 10.
```{r}
checkAlignment(dna_align[586:1170, 151:300], what = 1)
```

The end of the alignment for the first half of the sequences. Note that there does not seem to be any large gaps as in the previous alignments but there are still some smaller gaps that are visible.
```{r}
checkAlignment(dna_align[1:585, 301:422], what = 1)
```

The end of the alignment for the second half of the sequences. Note that there does not seem to be any large gaps as in the previous alignments but there are still some smaller gaps that are visible.
```{r}
checkAlignment(dna_align[586:1170, 301:422], what = 1)
```

The distribution of gap lengths. Note that there are over 8000 gaps of length 1.
```{r}
checkAlignment(dna_align, what = 2)
```

The shannon index (H) for each sequence position.
```{r}
checkAlignment(dna_align, what = 3)
```

The number of observed bases for each sequence position. Note that all positions have 1.0 base and most positions have 2.0 bases.
```{r}
checkAlignment(dna_align, what = 4)
```

We are going to remove the sequences with too many gaps. First we will visualize the gaps.
```{r}
seqLen <- as.numeric(lapply(dna_bin, length))
qplot(seqLen) +
  theme_bw()
```

From the graph above, we will use 290 as a cutoff and then visualize the alignment without those gaps.
```{r}
keepSeq <- seqLen > 290
dna_subset <- dna_align[keepSeq, ]
checkAlignment(dna_subset, what = 1)
```

Now we will redo the alignment without the sequences with the large gaps.
```{r}
dna_sub_align <- muscle(dna_subset, quiet = F)
```

## Analyze

We used the dist.dna function to compute a matrix of pairwise distances from DNA sequences using the k80 model of DNA evolution.
```{r}
dnaDM<-dist.dna(dna_sub_align, model="K80")
dnaDMmat<-as.matrix(dnaDM)
```

Next we generated a heat-map comparing the sequence similarities, however, to do this we had to utilize the melt function to turn our 'DM' file into a linear matrix which can then be plotted as a heat map. We then used ggplot to construct the heatmap according to certain specifications. 
```{r}
PDat<-melt(dnaDMmat)
dim(PDat)
ggplot(data = PDat, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile()+scale_fill_gradientn(colours=c("yellow","blue","green","red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

pdf(width=20,height=10, "HeatMap_B1285.pdf")
ggplot(data = PDat, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile()+scale_fill_gradientn(colours=c("yellow","blue","green","red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
dev.off()
```

## Phylogenetic Tree Building

To understand the evolutionary releationship among all 1170 samples, a phylogentic tree using the neighbour joining method was created. The distance matrix created earlier was inputted as the data. The neighbour joining method was used over the minimium evolution method becasuse each sequence read was relatively small (<300) while the number of taxa compared was high. The branch length was specified to 'none' to better focus on the relationship among the taxa rather than the evolutionary distance among them. We decided not to include tip labels because we are interested in the overall evolutionary relationship not the relationships between individual sequences. In addtion, not including the tip labels makes the tree more readable.
```{r}
dnaTree<-nj(dnaDM)
str(dnaTree)
ggtree(dnaTree, branch.length='none')

pdf(width=20,height=10, "PhylogeneticTree_B1285")
ggtree(dnaTree, branch.length='none')
dev.off()
```






















